### DISCUSSION TITLE
Segmentation fault 11 for addRedactAnnot and apply_redactions

### DISCUSSION THREAD

**Discussion by malizheng** [pymupdf/PyMuPDF#1173] [Upstream Bugs]
import fitz
cur_pdf = fitz.open("C_2010638-2.pdf") 
r=fitz.Rect([76.10399627685547, 239.55531311035156, 543.1917114257812, 254.18092346191406])
for cur_page in cur_pdf:
      cur_page.addRedactAnnot(r,fill=(1, 1, 1))
      cur_page.apply_redactions() 
cur_pdf.save("C_2010638-1.pdf",garbage=4, clean=1,deflate=True)

There is an error: Segmentation fault.

[C_2010638-2.pdf](https://github.com/pymupdf/PyMuPDF/files/5864755/C_2010638-2.pdf)

**Comment by JorjMcKie** [comment]
That's a bad one! Happens in MuPDF code - not PyMuPDF.
So I need to write a reproducer for MuPDF and submit to their issue tracker ...

**Comment by JorjMcKie** [comment]
Submitted a bug to MuPDF: https://bugs.ghostscript.com/show_bug.cgi?id=703388 for your reference.

**Comment by malizheng** [comment]
Thank you very much, I will pay attention to this.

**Comment by JorjMcKie** [comment]
Let us keep this open until resolved.

**Comment by JorjMcKie** [comment]
Update:
MuPDF was very fast in **_confirming_** this upstream bug. They also told me that it has already been resolved for the next version.

**Comment by malizheng** [comment]
Thank you, how to replace mupdf with the latest version in pymupdf? Or can you solve it by upgrading pymupdf?

**Comment by JorjMcKie** [comment]
Is that urgent for you?
The normal process would now be to wait for version 1.19.0 - meaning at least another 2 or 3 months. They have a bi-annual release schedule and 1.18.0 came out 2020-10-07.

**Comment by malizheng** [comment]
Yes, do you know what caused this error?  
I hope to know in advance that this error will occur, and skip this error without executing these two statements.

**Comment by JorjMcKie** [comment]
This error happens inside MuPDF under rare conditions. It has to with the special way the **_text_** is coded,
I have experimented a bit and rewritten that page's text using my font replacement tool. Then used your script with the rewritten page and it worked.

I guess I could give better recommendations if I knew what exactly you are trying to achieve. So maybe, just overlaying text with a white rectangle would be sufficient (i.e. not redacting it).

What is your operating system and Python version and bitness?

**Comment by malizheng** [comment]
The environment:  64-bit windows 10 and python3.7 .
Thank you very much. I used the PyPDF2 package to rewrite this pdf file. The generated file will not report this error, but this will require a local backup of the original file. Would you please provide your source code?

**Comment by JorjMcKie** [comment]
You mean the source code for the font replacement?
Here is a zip containing the pre-version of my next PyMuPDF and two scripts for font replacement. Please install.
[issue-866.zip](https://github.com/pymupdf/PyMuPDF/files/5878438/issue-866.zip)

Font replacement works in 3 steps:
1. run ``python repl-fontnames.py your.pdf`` - produces a JSON file
2. edit the produced JSON ``your.pdf-fontnames.json`` and replace each "keep" you want with the name of a new font
3. run ``python repl-font.py your.pdf`` produces the PDF ``your-new.pdf``

I also have included a JSON file I produced for the PDF you gave me. Maybe it is sufficient for running **_only step 3_** above.

**Comment by malizheng** [comment]
Thank you, I have received it.

**Comment by JorjMcKie** [comment]
I did a hack and simply copied the MuPDF modification into my local copy of MuPDF, then generated PyMuPDF anew ... and it seems to work!
Here is your wheel:
[PyMuPDF-1.18.7-cp37-cp37m-win_amd64.whl.zip](https://github.com/pymupdf/PyMuPDF/files/5878548/PyMuPDF-1.18.7-cp37-cp37m-win_amd64.whl.zip)

**Comment by JorjMcKie** [comment]
Did the last wheel resolve your issue?

**Comment by malizheng** [comment]
Thank you very much, it also solves my problem. Use the tool to generate a blank pdf in advance, and then overlay the original PDF on it, the program will not report an error.

**Comment by JorjMcKie** [comment]
Notes for others interested in this bug:
* it only happens in rare situations
* if needed, use a circumvention as presented by @malizheng in previous post
* the next MuPDF (and hence PyMuPDF) will contain a fix
* I included the fix in my local MuPDF 1.18.0 installation and will generate the wheels for v1.18.7 with it. So only those generating from sources will remain affected until the next MuPDF version (1.19.0 presumably).

**Comment by jmoreno11** [comment]
Hello @JorjMcKie, do you also happen to have the wheel for macOS? It's been happening to me for a while but I can't share sample documents because they contain sensitive information… It would be great if you had it, otherwise I'll have to wait for the next version, as you said.
Thank you in advance.
